ingest-app/src/cmr/ingest/services/granule_bulk_update_service.clj

```diff
@@ -1,79 +1,162 @@
-   ;; Queue the bulk update event
-   (handle-bulk-update-event context provider-id task-id bulk-update-params)
-   task-id))
+  "Validate the granule bulk update request, save rows to the db for task
+  and granule statuses, and queue bulk granule update. Return task id, which comes
+  from the db save."
+  [context provider-id json-body user-id]
+  (validate-granule-bulk-update json-body)
+  (let [instructions (-> json-body
+                         (json/parse-string true)
+                         request->instructions)
+        task-id (try
+                  (data-granule-bulk-update/create-granule-bulk-update-task
+                   context
+                   provider-id
+                   user-id
+                   json-body
+                   instructions)
+                  (catch Exception e
+                    (error "validate-and-save-bulk-granule-update caught exception:" e)
+                    (let [message (.getMessage e)
+                          user-facing-message (if (string/includes? message "GBUT_PN_I")
+                                                "Granule bulk update task name needs to be unique within the provider."
+                                                message)]
+                      (errors/throw-service-errors
+                       :invalid-data
+                       [(str "error creating granule bulk update task: " user-facing-message)]))))]
+    ;; Queue the granules bulk update event
+    (ingest-events/publish-gran-bulk-update-event
+     context
+     (ingest-events/granules-bulk-event provider-id task-id user-id instructions))
+    task-id))
+
+(defn handle-granules-bulk-event
+  "For each granule-ur, queueu granule bulk update messages"
+  [context provider-id task-id bulk-update-params user-id]
+  (doseq [instruction bulk-update-params]
+    (ingest-events/publish-gran-bulk-update-event
+     context
+     (ingest-events/ingest-granule-bulk-update-event provider-id task-id user-id instruction))))
+
+(defmulti add-opendap-url
+  "Add OPeNDAP url to the given granule concept."
+  (fn [context concept url]
+    (mt/format-key (:format concept))))
+
+(defmethod add-opendap-url :echo10
+  [context concept url]
+  (echo10/add-opendap-url (:metadata concept) url))
+
+(defmethod add-opendap-url :default
+  [context concept url]
+  (errors/throw-service-errors
+   :invalid-data (format "Add OPeNDAP url is not supported for format [%s]" (:format concept))))
+
+(defmulti update-granule-concept
+  "Perform the update of the granule concept."
+  (fn [context concept bulk-update-params user-id]
+    (keyword (:event-type bulk-update-params))))
+
+(defmethod update-granule-concept :UPDATE_FIELD:OPeNDAPLink
+  [context concept bulk-update-params user-id]
+  (let [{:keys [format metadata]} concept
+        {:keys [granule-ur url]} bulk-update-params
+        updated-metadata (add-opendap-url context concept url)]
+    (if-let [err-messages (:errors updated-metadata)]
```

Comment:
Readability nitpick
```suggestion
        updated-metadata (add-opendap-url context concept url)
        err-messages (:errors updated-metadata)]
    (if (not= (nil? err-messages))
```

Reply:
I think `if-let` is actually more readable. `(not= (nil? err-messages))` is wrong. It should have been `(not (nil? err-messages))` or `(seq err-messages)`.